### DISCUSSION TITLE
Datafetchers as completable futures don't run in parallel

### DISCUSSION THREAD

**Discussion by prograhammer** [Netflix/dgs-framework#509] [Q&A]
I created two simple datafetchers as completable futures but they don't run in parallel. What am I missing here?

Simple schema:
![image](https://user-images.githubusercontent.com/4474660/127939638-1896f998-eaa0-4cc6-a4de-51e583594931.png)

Users DataFetcher:

![image](https://user-images.githubusercontent.com/4474660/127939698-3c53cdf8-ead1-41a9-ba99-308c9887f93e.png)

Products DataFetcher:
![image](https://user-images.githubusercontent.com/4474660/127939731-8c509bbc-86f0-42ef-a214-1c1862bc4fa6.png)

They don't run in parallel:
![image](https://user-images.githubusercontent.com/4474660/127939755-dcd7e14f-3d76-486e-ad54-f77eaf4c51f9.png)

I could create the futures and join them within a single datafetcher but that defeats the nice annotation ability of DGS right? It is because one is Query and the other is an Object Type (on different levels, so-to-speak)?

**Comment by paulbakker** [comment]
You’ll need to provide an Executor as the second argument when creating the
CompletableFuture

**Reply by prograhammer** [reply]
Hi @paulbakker 

I updated to explicitly pass in an executor (to both datafetchers) but still they don't run in parallel:

```java
    Executor executor = Executors.newFixedThreadPool(10);

    return CompletableFuture.supplyAsync(() -> {
      try {
          System.out.println("User task execution started.");
          Thread.sleep(5000);
          System.out.println("User task execution stopped.");
      } catch (Exception e) {
          e.printStackTrace();
      }
      return users;
    }, executor);
```

Am I doing this wrong?

**Reply by prograhammer** [reply]
Oh I probably need to do this at startup and then use the same executor across datafetchers.....let me try that...

**Reply by prograhammer** [reply]
No this didn't help:

Executor bean:
```java
@Service
public class ExecutorService {

    private final Executor executor;

    @Autowired
    public ExecutorService() {
        this.executor = Executors.newFixedThreadPool(10);
    }

    public Executor getExecutor() {
        return this.executor;
    }
    
}
```

Updated to User (and I also updated Product) datafetcher:
```java
  @Autowired
  private ExecutorService executorService;

  @DgsData(parentType = "Query", field = "users")
  public CompletableFuture<List<User>> getUsers(DataFetchingEnvironment dfe) {

    String userId = dfe.getArgument("userId");

    return CompletableFuture.supplyAsync(() -> {
      try {
          System.out.println("User task execution started.");
          Thread.sleep(5000);
          System.out.println("User task execution stopped.");
      } catch (Exception e) {
          e.printStackTrace();
      }
      return users;
    }, executorService.getExecutor());
```

Still does not run in parallel.

**Comment by prograhammer** [comment]
Any example of how to use an executor @paulbakker to get datafetchers to actually work in parallel?

**Comment by paulbakker** [comment]
This was answered in #512:

> It took me a minute to figure out what's going on. The problem comes from the relation between `User` and `Product`. Because `products` is a field of `User`, it first needs to resolve the parent (`User` in this case) before it can continue to resolve subfields.
> Imagine for example that your `products` datafetcher would use the `getSource` method to get its parent (the `User`). That wouldn't work if `User` wasn't resolved yet.
> 
> If products and user were two unrelated fields on `Query`, your code would work. I also found that my suggestion about using an `Executor` isn't actually required.
> 
> We have a simple test in the example embedded in the framework's repository that demonstrates this: https://github.com/Netflix/dgs-framework/blob/master/graphql-dgs-example-shared/src/test/java/com/netflix/graphql/dgs/example/shared/ConcurrentDataFetcherTest.java
> 
> Of course, that doesn't answer the question how you _could_ make such a scenario parallel. I think the best option is a "look ahead" data fetcher as described here: https://netflix.github.io/dgs/advanced/context-passing/#pre-loading.
> 
> I'm closing this issue because I don't think it's further actionable, but feel free to ask more questions.
